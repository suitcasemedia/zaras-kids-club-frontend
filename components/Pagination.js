import Head from 'next/head';
import Link from 'next/dist/client/link';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import PaginationStyles from './styles/PaginationStyles';
import DisplayError from './ErrorMessage';
import { perPage } from '../config';

export const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    _allProductsMeta {
      count
    }
  }
`;
export default function Pagination({ page }) {
  const { error, loading, data } = useQuery(PAGINATION_QUERY);

  if (loading) return <p>loading....</p>;
  if (error) return <DisplayError error={error} />;

  const { count } = data?._allProductsMeta;
  const pageCount = Math.ceil(count / perPage);
  return (
    <PaginationStyles>
      <Head>
        <title>Kids CLub - page {page} of __</title>
      </Head>

      <Link href={`/products/${page - 1}`}>
        <a aria-disabled={page <= 1}>← Prev</a>
      </Link>
      <p>
        Page {page} of {perPage}{' '}
      </p>
      <p>{count} Items Total </p>
      <Link href={`/products/${page + 1}`}>
        <a aria-disabled={page >= pageCount}>Next →</a>
      </Link>
    </PaginationStyles>
  );
}
