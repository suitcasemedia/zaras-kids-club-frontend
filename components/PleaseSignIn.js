import { PropTypes } from 'prop-types';
import { useUser } from './User';
import SignIn from './SignIn';

export default function GatedSignIn({ children }) {
  const me = useUser();
  if (!me) return <SignIn />;
  return children;
}

GatedSignIn.propTypes = {
  children: PropTypes.object,
};
