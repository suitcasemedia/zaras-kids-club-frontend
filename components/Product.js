import PropTypes from 'prop-types';
import Link from 'next/link';
import ItemStyles from './styles/ItemStyles';
import Title from './styles/Title';
import PriceTag from './styles/PriceTag';
import formatMoney from '../lib/formatMoney';
import DeleteProduct from './deleteProduct';
import AddToCart from './AddToCart';

export default function Product({ product }) {
  return (
    <ItemStyles>
      <img
        alt={product.name}
        src={product?.photo?.image?.publicUrlTransformed}
      />
      <Title>
        <Link href={`/product/${product.id}`}>{product.name}</Link>
      </Title>
      <PriceTag>{formatMoney(product.price)}</PriceTag>
      <p>{product.description}</p>
      <div className="buttonList">
        <Link
          href={{
            pathname: 'update',
            query: {
              id: product.id,
            },
          }}
        >
          Edit
        </Link>
        <AddToCart id={product.id} />
        <DeleteProduct id={product.id}>Delete</DeleteProduct>
      </div>
    </ItemStyles>
  );
}
Product.propTypes = {
  product: PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.string,
    description: PropTypes.string,
    price: PropTypes.number,
    photo: PropTypes.shape({
      id: PropTypes.string,
      image: PropTypes.shape({
        publicUrlTransformed: PropTypes.string,
      }),
    }),
  }),
};
