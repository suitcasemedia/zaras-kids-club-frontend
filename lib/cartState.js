import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;
function CartStateProvider({ children }) {
  // this is its own custom provider

  const [cartOpen, setCartOpen] = useState(false);
  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// make a custom hook for acessing the cart local state

function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}
export { CartStateProvider, useCart };
