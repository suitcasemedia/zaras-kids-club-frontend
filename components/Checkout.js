import { loadStripe } from '@stripe/stripe-js';
import styled from 'styled-components';
import {
  CardElement,
  Elements,
  useElements,
  useStripe,
} from '@stripe/react-stripe-js';
import { useState } from 'react';
import nProgress from 'nprogress';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { useRouter } from 'next/dist/client/router';
import { useCallback } from 'react/cjs/react.production.min';
import SickButton from './styles/SickButton';
import { useCart } from '../lib/cartState';
import { CURRENT_USER_QUERY } from './User';

const CheckoutFormStyles = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.04);
  border: 1px solid rgba(0, 0, 0, 0.06);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
`;

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id
        name
      }
    }
  }
`;
function CheckoutForm() {
  const { closeCart } = useCart();
  const [error, setError] = useState('');
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();
  const router = useRouter();
  const [checkout, { error: graphQlError }] = useMutation(
    CREATE_ORDER_MUTATION,
    {
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );
  async function handleSubmit(e) {
    // stoop form from loading turn loader on
    e.preventDefault(e);
    setLoading(true);
    /// 2 start page transition
    nProgress.start();
    // 3 create the payment methos via sstripe

    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: 'card',
      card: elements.getElement(CardElement),
    });

    console.log(paymentMethod);
    // 4 handle any errors from stripe

    if (error) {
      setError(error);
      nProgress.done();
      return; // stops checkout happening
    }
    // 5. send the tokenfrom step 3 to our keystone server via custom mutation

    const order = await checkout({
      variables: {
        token: paymentMethod.id,
      },
    });
    console.log(`finished with the order!`);
    console.log({ order });
    // 6 change the page to view order
    router.push({
      pathname: '/order/[id]',
      query: { id: order.data.checkout.id },
    });
    // 7 close the cart

    closeCart();
    // 8 turn the loader off
    setLoading(false);
    nProgress.done();
  }

  return (
    <CheckoutFormStyles onSubmit={handleSubmit}>
      {error && <p style={{ fontSize: 14 }}>{error.message}</p>}
      {graphQlError && <p style={{ fontSize: 14 }}>{graphQlError.message}</p>}
      <CardElement />
      <SickButton>Check Out Now</SickButton>
    </CheckoutFormStyles>
  );
}

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

export default function Checkout() {
  return (
    <Elements stripe={stripeLib}>
      <CheckoutForm />
    </Elements>
  );
}
