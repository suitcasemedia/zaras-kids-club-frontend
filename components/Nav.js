import Link from 'next/link';
import NavStyles from './styles/NavStyles';
import { useUser } from './User';
import SignOut from './SignOut';
import { useCart } from '../lib/cartState';
import CartCount from './CartCount';
import Cart from './Cart';

export default function Nav() {
  const user = useUser();
  const { openCart } = useCart();

  return (
    <NavStyles>
      {/* <Link href="/products">Shop</Link> */}
      {user && (
        <>
          {/* <Link href="/sell">Sell</Link>

          <Link href="/orders">Orders</Link>
          <Link href="/account">Account</Link> */}
          <SignOut />
          <button onClick={openCart}>
            My Cart
            <CartCount
              count={user.cart.reduce(
                (tally, cartItem) =>
                  tally + (cartItem.product ? cartItem.quantity : 0),
                0
              )}
            />
          </button>
        </>
      )}
      <Link href="/about">About</Link>
      <Link href="/ourteam">Our Team</Link>

      {/* {!user && (
        <>
          <Link href="/signin">Sign in</Link>
        </>
      )} */}
    </NavStyles>
  );
}
