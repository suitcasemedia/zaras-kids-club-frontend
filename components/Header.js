import Link from 'next/link';
import styled from 'styled-components';
import Nav from './Nav';
import Cart from './Cart';
import Search from './Search';
import Logo from './styles/Logo';
import LogoWrapper from './styles/LogoWrapper';

const HeaderStyles = styled.header`
  background-color: var(--purple);
  display: flex;
  // flex-wrap: wrap; /* this */
  justify-content: space-between;
  > div {
    display: flex;
    flex-wrap: 400px;
  }

  .bar {
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: stretch;
  }
  .sub-bar {
    display: grid;
    grid-template-columns: 1fr auto;

    color: var(--purple);
    padding: 20px;
    width: 400px;
  }
`;
export default function Header() {
  return (
    <HeaderStyles>
      <div>
        <div className="bar">
          <LogoWrapper>
            <Logo>
              {' '}
              <Link href="/">Zara's Kids Club</Link>
            </Logo>
          </LogoWrapper>
        </div>
      </div>
      <div>
        <Nav />
        {/* <Cart /> */}
      </div>

      {/* <div>
        <div className="sub-bar">
          <Search />
        </div>
        <div>
        
        </div>
      </div> */}
    </HeaderStyles>
  );
}
