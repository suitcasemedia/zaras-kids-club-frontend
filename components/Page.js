import PropTypes from 'prop-types';
import styled, { createGlobalStyle } from 'styled-components';
import Header from './Header';

const GLobalStyles = createGlobalStyle`
@font-face {
  font-family: 'radnika_next';
  src: url('/static/radnikanext-medium-webfont.woff2') format('woff2');
  font-weight: normal;
  font-style: normal;
}
@font-face {
  font-family: 'peaches_and_cream';
  src: url('/static/peaches-and-cream.woff') format('woff');
  font-weight: normal;
  font-style: normal;
}
  html {
    --red: #ff0000;
    --black : #393939;
    --grey: #3A3A3A;
    --gray: var(--grey);
    --lightGrey: #e1e1e1;
    --lightGray: var(--lightGrey);
    --offWhite: #ededed;
    --maxWidth: 1000px;
    --bs: 0 12px 24px 0 rgba(0,0,0,0.09);
    --purple: rgb(86,57,150);
    --yellow: #FFEE00;
    --pink: #FF99B3;
    --light-purple: #C07EEC ;
    --off-white: rgba(255, 255, 255, 0.9);
    box-sizing: border-box;
    font-size: 10px;
    background: rgba(255, 238,0,0.6)
    

  }
  *,*:before, *:after{
    box-sizing: inherit;
  }
  body{
    font-family: 'radnika_next', system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
    padding: 0;
    margin: 0;
    line-height: 2;
    color: var(--purple);
  }
  a{
    text-decoration: none;
    color: var(--black);
  }
  a:hover{
    text-decoration: underline;
  }
  button{
    font-family: 'radnika_next' system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
   

  }
  `;

const InnerStyles = styled.div`
  max-width: var(--maxWidth);
  margin: 0 auto;
  padding: 2rem;
`;

export default function Page({ children }) {
  return (
    <div>
      <GLobalStyles />
      <Header />
      <InnerStyles>{children}</InnerStyles>
    </div>
  );
}

Page.propTypes = {
  //   children: PropTypes.oneOf([
  //     PropTypes.arrayOf(PropTypes.node),
  //     PropTypes.node,
  //   ]),
  children: PropTypes.any,
};
