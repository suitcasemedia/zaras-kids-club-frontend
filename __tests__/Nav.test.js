import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { ApolloClient } from '@apollo/client';
import { de } from 'date-fns/locale';
import Nav from '../components/Nav';
import { fakeUser, fakeCartItem } from '../lib/testUtils';
import { CartStateProvider } from '../lib/cartState';
import { CURRENT_USER_QUERY } from '../components/User';

// Make some mocks for being logged in and not logging in
// and logged in with cart items
const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: null } },
  },
];
const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: fakeUser() } },
  },
];

const signedInWithCartItemsMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        authenticatedItem: fakeUser({
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
        }),
      },
    },
  },
];

console.log(signedInWithCartItemsMocks);
describe('<Nav/>', () => {
  it('renders a minimal nav when signed out', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={notSignedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );

    debug();
    expect(container).toHaveTextContent('Sign in');
    // q: whats the difference between screen and container?
    // a: screen is a wrapper around container
    expect(container).toMatchSnapshot();
    const link = screen.getByText('Sign in');
    expect(link).toHaveAttribute('href', '/signin');
  });

  it('renders full nav when signed in', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={signedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByText('Account');
    debug();
  });
});
