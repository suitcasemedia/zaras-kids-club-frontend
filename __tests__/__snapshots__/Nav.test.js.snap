// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a minimal nav when signed out 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 jsRUpl"
  >
    <a
      href="/products"
    >
      Shop
    </a>
    <a
      href="/signin"
    >
      Sign in
    </a>
  </ul>
</div>
`;
